waggle-dance-core/src/test/java/com/hotels/bdp/waggledance/client/TunnelingMetastoreClientBuilderTest.java

```diff
@@ -0,0 +1,138 @@
+import static org.mockito.Matchers.anyInt;
+import static org.mockito.Matchers.anyString;
+import static org.mockito.Matchers.eq;
+import static org.mockito.Mockito.doNothing;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+import org.apache.hadoop.hive.conf.HiveConf;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.mockito.Spy;
+import org.powermock.api.mockito.PowerMockito;
+import org.powermock.core.classloader.annotations.PrepareForTest;
+import org.powermock.modules.junit4.PowerMockRunner;
+
+import com.pastdev.jsch.tunnel.Tunnel;
+import com.pastdev.jsch.tunnel.TunnelConnectionManager;
+
+@RunWith(PowerMockRunner.class)
+@PrepareForTest(TunnelingMetastoreClientBuilder.class)
+public class TunnelingMetastoreClientBuilderTest {
+  private @Spy TunnelingMetastoreClientBuilder tunnelingMetastoreClientBuilder = new TunnelingMetastoreClientBuilder();
+  private @Mock TunnelConnectionManagerFactory tunnelConnectionManagerFactory;
+  private @Spy HiveConf hiveConf = new HiveConf();
+  private @Mock TunnelConnectionManager tunnelConnectionManager;
+  private @Mock Tunnel tunnel;
+
+  @Test
+  public void build() throws Exception {
+    hiveConf.set(WaggleDanceHiveConfVars.SSH_ROUTE.varname, "user@ec2-12-345-678-91.compute-1.amazonaws.com");
+    hiveConf.set(WaggleDanceHiveConfVars.SSH_PRIVATE_KEYS.varname, "private_key");
+    String originalMetastoreUri = "thrift://internal-test-shared-hive-metastore-elb-1234567891.us-west-1.elb.amazonaws.com:1234";
+    hiveConf.setVar(HiveConf.ConfVars.METASTOREURIS, originalMetastoreUri);
+
+    when(tunnelConnectionManagerFactory.create(anyString(), anyString(), anyInt(), anyString(), anyInt())).thenReturn(
+        tunnelConnectionManager);
+    when(tunnelConnectionManager.getTunnel(anyString(), anyInt())).thenReturn(tunnel);
+    doNothing().when(tunnelConnectionManager).open();
+
+    tunnelingMetastoreClientBuilder
+        .setHiveConf(hiveConf)
+        .setName("name")
+        .setReconnectionRetries(10)
+        .setTunnelConnectionManagerFactory(tunnelConnectionManagerFactory)
+        .setRemoteHost("remote")
+        .setSSHRoute("route")
+        .setLocalHost("local")
+        .setRemotePort(9083)
+        .build();
+
+    verify(tunnelConnectionManager, times(2)).getTunnel(anyString(), anyInt());
+    verify(tunnelConnectionManager).open();
+
+    //Check that Client is being created from a hiveConf different to that passed into the builder
+    PowerMockito.verifyPrivate(tunnelingMetastoreClientBuilder, times(0))
+                .invoke("clientFromLocalHiveConf", tunnelConnectionManager, hiveConf);
+    PowerMockito.verifyPrivate(tunnelingMetastoreClientBuilder, times(1))
+                .invoke("clientFromLocalHiveConf", eq(tunnelConnectionManager), any(HiveConf.class));
+  }
+
+  @Test
+  public void setHiveConf() throws Exception {
```

Comment:
we don't want to tests these methods but we do want to test the final result of `build()`